doc/PROPOSAL.md

```diff
@@ -0,0 +1,163 @@
+# Seguro
+
+A production wrapper around FoundationDB, ready for integration with `vere`.
+
+## Background
+
+The results of Phase 1 of the Seguro project demonstrate that FoundationDB is
+the best choice of distributed databases to utilize for the project. This is
+Phase 2 of the Seguro project, which implements an MVP event reader/writer
+module using FoundationDB as the underlying data store.
+
+See below for additional context:
+
+- [Seguro Prototype (Phase 1) Grant](https://urbit.org/grants/seguro-prototype)
+- [Seguro Prototype (Phase 1) Repository](https://github.com/wexpertsystems/seguro)
+- [Seguro & Armada Paper](https://gist.github.com/wexpert/0485a722185d5ee70742570036faf32f)
+
+## Objectives
+
+Phase 2 of the Seguro project can be considered as "all work up to actual
+integration with `vere`". Limiting Phase 2 to this is desirable because it i.)
+keeps the grant's scope of work at a reasonable level, and ii.) it allows more
+time for the following three ancillary runtime projects/considerations to
+achieve resolution:
+
+1. Asynchronous reads in `vere`
+2. Merging/releasing event log truncation
+   [urbit/urbit#7501](https://github.com/urbit/urbit/pull/5701)
+3. A decision on whether or not to add another process to the overall Urbit
+   runtime architecture
+
+Each of these three workstreams need to be resolved before Seguro can be
+integrated with `vere`.
+
+### Event Fragmentation Scheme
+
+FoundationDB supports a maximum value size of 100KB and recommends limiting
+value size to only 10KB for best performance. Since our data model requires
+arbitrarily sized byte buffer values for events, an event fragmentation
+algorithm is required to support events with larger payloads.
+
+Events which exceed the size limit of a single database value must be fragmented
+so that such events are stored across multiple database key/value pairs. To
+achieve event fragmentation, key/value pairs will be formatted like so:
+
+#### Keys/Values
+
+|           | encoding                                                                           | format                         |
+|-----------|------------------------------------------------------------------------------------|--------------------------------|
+| **key**   | `0x00` &#x7c; 64-bit event id (big-endian) &#x7c; 32-bit fragment id (big-endian)  | `0x00112233445566778899AABBCC` |
+| **value** | raw bytes starting at index `(i * <optimal fragment size>)` for fragment w/ id `i` | `0xdeadbeef...`                |
+
+Notes:
+- `|` = concatenation
+- The event ID and fragment ID are both contiguous, non-negative integers
+- The optimal fragment size, as documented above, will almost certainly be
+  `10 KB` = `10,000 bytes`
+- The first fragment (fragment `0`) will have a slightly different encoding for
+  the value, as documented below
+
+#### Fragment 0
```

Comment:
Will you put the reasoning behind using a different encoding for Fragment 0 here please?

Reply:
The reasoning is already included, check line 68